pkg/resourceinterpreter/customized/declarative/luavm/lua_convert.go

```diff
@@ -47,50 +49,40 @@ func ConvertLuaResultInto(luaResult lua.LValue, obj interface{}) error {
 	// ReplicaRequirements object.
 	//
 	// Here we handle it as follows:
-	//   1. Walk the object (lua table), delete the key with empty value (`nodeClaim` in this example):
-	//     {
-	//         resourceRequest: {
-	//             cpu: "100m"
-	//         }
-	//     }
-	//   2. Encode the object with luajson to be:
-	//     {"resourceRequest": {"cpu": "100m"}}
-	//   4. Finally, unmarshal the new json to object, get
+	//   1. Encode the object with luajson to be:
+	//     {"nodeClaim": [], "resourceRequest": {"cpu": "100m"}}
+	//   2. replace the empty value from `[]` to `{}` (`nodeClaim` in this example)
+	//     {"nodeClaim": {}, "resourceRequest": {"cpu": "100m"}}
+	//   3. Finally, unmarshal the new json to object, get
 	//     {
+	//         nodeClaim: {},
 	//         resourceRequest: {
 	//             cpu: "100m"
 	//         }
 	//     }
-	isEmptyDic := func(v *lua.LTable) bool {
-		count := 0
-		v.ForEach(func(lua.LValue, lua.LValue) {
-			count++
-		})
-		return count == 0
-	}
-
-	var walkValue func(v lua.LValue)
-	walkValue = func(v lua.LValue) {
-		if t, ok := v.(*lua.LTable); ok {
-			t.ForEach(func(key lua.LValue, value lua.LValue) {
-				if tt, ok := value.(*lua.LTable); ok {
-					if isEmptyDic(tt) {
-						// set nil to delete key
-						t.RawSetH(key, lua.LNil)
-					} else {
-						walkValue(value)
-					}
-				}
-			})
-		}
-	}
-	walkValue(luaResult)
-
+	//
+	//  Notes: don't worry that the value of a field is `[]`, and it originally represents an empty slice.
+	//  Because this situation will not appear in the following `jsonBytes`.
+	//  Supposing the value of a field is originally an empty slice:
+	//    1. if this field is omitempty, the key and value will both been removed in `jsonBytes`.
+	//    2. if this field is not omitempty, but required, empty slice is impossible to be created.
+	//    3. if this field is not omitempty, but optional, it has been verified also be removed in `jsonBytes`, e.g: `Rules` field in `Role` Type.
 	jsonBytes, err := luajson.Encode(luaResult)
 	if err != nil {
 		return fmt.Errorf("json Encode obj eroor %v", err)
 	}
 
+	// Only if `[]` is the value of a field, it will be converted to `{}`,
+	// otherwise, for example, if `[]` is a substring in a string value, it needs to be ignored
+	//
+	// e.g: assuming the given jsonBytes is: {"one-field":[],"one-label":"\"hello\":[]"}
+	// it is expected to convert to: {"one-field":{},"one-label":"\"hello\":[]"}
+	re, _ := regexp.Compile(`"([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]":\[]`)
```

Comment:
I'm a little skeptical about this regexp. If ```"hello":[]``` is used as the value of a field, it can actually match this regexp. etc, ```"{\"one-field\":`\"hello\":[]`}"```.
```go
func main() {
	re, _ := regexp.Compile(`"([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]":\[]`)
	fmt.Println(re.MatchString("{\"one-field\":`\"hello\":[]`}"))
}
```
```shell
PS C:\Users\Administrator\GolandProjects\awesomeProject> go run main.go
true
```

Reply:
never mind, the above example string is formatted incorrectly, the correct one should be `{"one_filed":"\"hello\":[]"}`. It is verified that the match fails, as expected

Reply:
thanks all the same~